### PR TITLE
[LT] Support Tensor.is_alias_of

### PR DESCRIPTION
**Pull request by alanwaketan**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #75350
* #75344
* #75247
* __->__ #75246

Summary:
Tensor.is_alias_of relies on Storage to perform. However, LTCTensorImpl was
not implemented with that in mind. This commit adds a fake storage to LazyTensor
as a marker to mark LazyTensors that point to the same storage. The reason
why it's not done at LTCTensorImpl is that LazyTensor maintains the view ops/alias
logic in LazyTensor class instead of relying on TensorImpl to do the check.

Test Plan:
./build/bin/test_lazy --gtest_filter=LazyOpsTest.IsAliasOf

Differential Revision: [D35389123](https://our.internmc.facebook.com/intern/diff/D35389123)

### CODE REVIEW

**Comment by alanwaketan** [comment]
@alanwaketan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35389123).

**Comment by alanwaketan** [comment]
Please refer to the #71378 for the discussion on why we support is_alias_of in this way instead of making the method virtual.

File: torch/csrc/lazy/core/tensor.cpp [hunk 1]
```diff
@@ -345,7 +347,9 @@ std::shared_ptr<LazyView> LazyTensor::CreateView(ViewInfo view_info) const {
 }
 
 LazyTensorPtr LazyTensor::CreateViewTensor(ViewInfo view_info) const {
-  return Create(CreateView(std::move(view_info)), GetDevice());
+  auto new_tensor = Create(CreateView(std::move(view_info)), GetDevice());
+  new_tensor->storage_ = Storage();
```
Comment (wconstab) [inline torch/csrc/lazy/core/tensor.cpp:351]: hmm, so we explicitly copy the Storage from one LT to another one if we know we are creating a view.

i was actually thinking we could tie it into the 'data()' object in some way, so if 2 LazyTensors happen to point to the same IR value then they also share the same storage.  I'm not actually sure if that is a desirable property, and maybe your approach is good enough.  It certainly seems to be simpler, which is nice.

**Comment by wconstab** [comment]
I think this change would be OK, but i'll wait for @bdhirsh to stamp

File: torch/csrc/lazy/core/tensor.cpp (hunk 1 above)
Comment (bdhirsh) [reply torch/csrc/lazy/core/tensor.cpp:351]: > if 2 LazyTensors happen to point to the same IR value then they also share the same storage

just curious - are you saying it's possible to be in a state where you have two LTC tensor's that aren't views of each other, but they actually point to the same IR under the hood?

**Review by bdhirsh — APPROVED** [review]
Looks ok to me! I'm hoping to have a mini prototype of integration functionalization into LTC pretty soon (where a bunch more changes are needed / stuff breaks, but hopefully I can show what the integration would look like at a high level)

**Comment by sujoysaraswati** [comment]
I have a question on the LazyTensor data_ptr() access. Since the has_storage() and storage() methods on LazyTensor would now return true and the fake storage, how does it affect APIs like data_ptr() for LazyTensor, is it planned to be supported for lazy tensor?

**Comment by sujoysaraswati** [comment]
> Please refer to the #71378 for the discussion on why we support is_alias_of in this way instead of making the method virtual. cc @sujoysaraswati.

Thanks @alanwaketan, we have followed the suggestion to use is_alias_of with a dummy storage and have closed #71378 now.

**Comment by alanwaketan** [comment]
> I have a question on the LazyTensor data_ptr() access. Since the has_storage() and storage() methods on LazyTensor would now return true and the fake storage, how does it affect APIs like data_ptr() for LazyTensor, is it planned to be supported for lazy tensor?

It shouldn't affect any exist APIs in the LazyTensor class. Or do you mean data_ptr() APIs somewhere else?

**Comment by sujoysaraswati** [comment]
> > I have a question on the LazyTensor data_ptr() access. Since the has_storage() and storage() methods on LazyTensor would now return true and the fake storage, how does it affect APIs like data_ptr() for LazyTensor, is it planned to be supported for lazy tensor?
> 
> It shouldn't affect any exist APIs in the LazyTensor class. Or do you mean data_ptr() APIs somewhere else?

My question is If we call a tensor.data_ptr() on a LazyTensor, will it work and provide a valid pointer?

**Comment by alanwaketan** [comment]
> > > I have a question on the LazyTensor data_ptr() access. Since the has_storage() and storage() methods on LazyTensor would now return true and the fake storage, how does it affect APIs like data_ptr() for LazyTensor, is it planned to be supported for lazy tensor?
> > 
> > 
> > It shouldn't affect any exist APIs in the LazyTensor class. Or do you mean data_ptr() APIs somewhere else?
> 
> My question is If we call a tensor.data_ptr() on a LazyTensor, will it work and provide a valid pointer?

It will return the empty DataPtr that is created in https://github.com/pytorch/pytorch/pull/75246/files#diff-824247c5519c382811fb78f1272745e2f1dfbce4fb85fbb5bb04f550ea836206R83.

Not sure about the implications on other functions built on top of the DataPtr API. Before we will throw an assertion which basically means LazyTensor won't touch any of these APIs. And the for the future, we will deprecate this and use functionalization. Maybe worth a TORCH_WARN, but I don't see any particular issues here. @bdhirsh What do you think?

**Comment by sujoysaraswati** [comment]
> 

Yes, I was referring to the empty data pointer value that would be returned and if this would create issue for any caller code looking for a valid data pointer. Could you please also give some details on how this is planned to change with functionalization?

**Comment by bdhirsh** [comment]
That's true that accessing the data ptr won't give you a nice error message anymore.

With functionalization (integration with LTC is underway), the lazy tensors will be wrapped in a `[FunctionalTensorWrapper](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/FunctionalTensorWrapper.h)` - it also has a null data ptr, but it can do more than just track aliases - it properly propagates mutations across aliases, and tracks sizes/stride information.

**Comment by alanwaketan** [comment]
@alanwaketan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35389123).

**Pull request closed without merging**